aptos-move/aptos-vm/src/sharded_block_executor/sharded_aggregator_service.rs

```diff
@@ -0,0 +1,155 @@
+    state_store::state_key::StateKey, transaction::TransactionOutput,
+    write_set::TOTAL_SUPPLY_STATE_KEY,
+};
+use rayon::prelude::*;
+use serde::de::DeserializeOwned;
+use std::ops;
+
+pub fn get_state_value<S: StateView, T: DeserializeOwned>(
+    state_key: &StateKey,
+    state_view: &S,
+) -> Option<T> {
+    let value = state_view
+        .get_state_value_bytes(state_key)
+        .ok()?
+        .map(move |value| bcs::from_bytes(value.as_slice()));
+    value.transpose().map_err(anyhow::Error::msg).unwrap()
+}
+
+/// This class ensures that deltas can use all 128 bits without having to let go of the sign bit for
+/// cases where the delta is negative. That is, we don't have to use conversions to i128.
+/// However, it does not handle overflow and underflow. That is, it will indicate to the caller of
+/// the faulty logic with their usage of deltas.
+#[derive(Clone, Copy)]
+struct DeltaU128 {
+    delta: u128,
+    is_positive: bool,
+}
+
+impl DeltaU128 {
+    pub fn get_delta(minuend: u128, subtrahend: u128) -> Self {
+        if minuend >= subtrahend {
+            Self {
+                delta: minuend - subtrahend,
+                is_positive: true,
+            }
+        } else {
+            Self {
+                delta: subtrahend - minuend,
+                is_positive: false,
+            }
+        }
+    }
+
+    fn add_delta(self, other: u128) -> u128 {
+        if self.is_positive {
+            self.delta + other
+        } else {
+            other - self.delta
+        }
+    }
+}
+
+impl Default for DeltaU128 {
+    fn default() -> Self {
+        Self {
+            delta: 0,
+            is_positive: true,
+        }
+    }
+}
+
+impl ops::Add for DeltaU128 {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        if self.is_positive == rhs.is_positive {
+            Self {
+                delta: self.delta + rhs.delta,
+                is_positive: self.is_positive,
+            }
+        } else {
+            let (pos, neg) = if self.is_positive {
+                (self.delta, rhs.delta)
+            } else {
+                (rhs.delta, self.delta)
+            };
+            if pos >= neg {
+                Self {
+                    delta: pos - neg,
+                    is_positive: true,
+                }
+            } else {
+                Self {
+                    delta: neg - pos,
+                    is_positive: false,
+                }
+            }
+        }
+    }
+}
+
+pub fn aggregate_and_update_total_supply<S: StateView>(
+    sharded_output: &mut Vec<Vec<Vec<TransactionOutput>>>,
+    global_output: &mut [TransactionOutput],
+    state_view: &S,
+) {
+    let num_shards = sharded_output.len();
+    let num_rounds = sharded_output[0].len();
+    let total_supply_base_val: u128 = get_state_value(&TOTAL_SUPPLY_STATE_KEY, state_view).unwrap();
+
+    let mut aggr_total_supply_delta = vec![DeltaU128::default(); num_shards * num_rounds + 1];
```

Comment:
why do you need a +1 here?

Reply:
First value is 0 which is delta for shard 0;
+1 is the delta for global shard

Reply:
Let's add a comment here for the same :)

Reply:
done

Reply:
done